Add http metrics

francoisferrand opened the pull request:
Issue: CLDSRV-63

francoisferrand committed:
Tweak http metrics

* Remove `cloud_server` prefix from standard metrics
* Remove `cloud_server` label
* Add a route label, to store the operation/api method

Issue: CLDSRV-63

francoisferrand committed:
Add http requests duration metrics

Issue: CLDSRV-63

francoisferrand committed:
Fix error handling on prometheus endpoint

Issue: CLDSRV-63

francoisferrand committed:
Add http active requests metrics

Issue: CLDSRV-63

francoisferrand committed:
Add unit tests

Issue: CLDSRV-63

francoisferrand committed:
Update prom-client to 14.0.1

This is needed to support 'optional' labels.

Issue: CLDSRV-63

francoisferrand committed:
Fix counting of http requests

All requests must be counted, includes rejected.

Issue: CLDSRV-63

francoisferrand committed:
Add functionnal test

Issue: CLDSRV-63

francoisferrand committed:
Fix missing response on malformed requests

The server was not replying to a request (missing "resourceType") made
on backbeat or workflow-engine-operator internal routes.

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/cloudserver/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-63 contains:


* `8.3.1`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.3.2`


Please check the `Fix Version/s` of CLDSRV-63, or the target
branch of this pull request.

miniscruff commented on tests/functional/report/monitoring.js:
```diff
@@ -0,0 +1,106 @@
+            http.get({ host: conf.ipAddress, path: '/metrics', port: 8002 }, res => {
+                assert.strictEqual(res.statusCode, 200);
+
+                let body = '';
+                res.on('data', chunk => { body += chunk; });
+                res.on('end', () => resolve(body));
```
Does not really matter much for tests but I recommend using an array and push, with a join in end. Slightly more efficient. We also use this pattern a lot so it would be consistent. 

```suggestion
                let body = [];
                res.on('data', chunk => { body.push(chunk); });
                res.on('end', () => resolve(body.join('')));
```

miniscruff commented on tests/functional/report/monitoring.js:
```diff
@@ -0,0 +1,106 @@
+    function parseMetric(metrics, name, labels) {
+        const labelsString = Object.entries(labels !== undefined ? labels : {
+            method: 'GET', code: '200', route: 'healthcheck',
+        }).map(e => `${e[0]}="${e[1]}"`).join(',');
```
```suggestion
        }).map(e => e.join('=')).join(',');
```

francoisferrand replied:
This does not work, we need to add the double-quotes in the string

miniscruff commented on tests/functional/report/monitoring.js:
```diff
@@ -0,0 +1,106 @@
+        ['/_/metadata',     { method: 'GET', code: '403', route: 'routeMetadata' }],
+        ['/_/workflow-engine-operator',
+                            { method: 'GET', code: '405', route: 'routeWorkflowEngineOperator' }],
+    ].forEach(testcase => {
+        const path = testcase[0];
+        const labels = testcase[1];
```
Does this work in node 10?
```suggestion
    ].forEach([path, labels] => {
```

francoisferrand replied:
if you also put parenthesis `([path, labels])`, then it works indeed.
thanks!

miniscruff commented on tests/functional/report/monitoring.js:
```diff
@@ -0,0 +1,106 @@
+    ].forEach(testcase => {
+        const path = testcase[0];
+        const labels = testcase[1];
+        it(`it should count http ${labels.method} requests metrics on ${path}`, async () => {
```
You don't need to include it in descriptions.
```suggestion
        it(`should count http ${labels.method} requests metrics on ${path}`, async () => {
```

miniscruff commented on tests/unit/utils/monitoring.js:
```diff
@@ -0,0 +1,69 @@
+
+    it('it should return some metrics', async () => {
+        await fetchMetrics({ method: 'GET', url: '/metrics' }, res);
+        assert(res.writeHead.calledOnceWith(200) || !res.writeHead.called);
```
using `||` for assertions feels weird, we should know which one it is

francoisferrand replied:
i agree, though i did this to match the behavior of http module... (and thus not depend on the actual implementation).
But I fixed the code to explicitely call `writeHead`, so no reason to keep it indeed.

miniscruff commented on tests/unit/utils/monitoring.js:
```diff
@@ -0,0 +1,69 @@
+        assert(res.end.args[0][0].includes('\nnodejs_active_handles_total '));
+    });
+
+    it('it should have http duration histogram metrics', async () => {
```
I feel like these last 3 tests can be another table test

francoisferrand replied:
i fear this may increase the tests duration & flakiness...

williamlardier commented on lib/utilities/monitoringHandler.js:
```diff
@@ -167,23 +177,23 @@ function routeHandler(req, res, log, cb) {
  * @return {void}
  */
-function monitoringHandler(clientIP, req, res, log) {
+async function monitoringHandler(clientIP, req, res, log) {
```
If we don't await any of the returned/called function, why do we want this function to be async?

francoisferrand replied:
it calls `routeHandler()`, which is async because it calls/awaits' promClient's `metrics()` function...
but it cannot be awaited anyway, i'll just try to remote it.

williamlardier replied:
Unless you are awaiting an async function, you should be able to call async function from sync ones, it will behave as expected :smile:

williamlardier reviewed (commented):
Quick review

francoisferrand committed:
fixup! Update prom-client to 14.0.1

francoisferrand committed:
fixup! Add functionnal test

francoisferrand committed:
fixup! Add functionnal test

lavigne958 commented on lib/utilities/monitoringHandler.js:
```diff
@@ -44,6 +44,10 @@ const httpRequestDurationSeconds = new client.Histogram({
 });
+const httpActiveRequests = new client.Gauge({
+    name: 'http_active_requests',
+    help: 'Total number of HTTP requests',
```
This is miss-leading, it should be something that explicitly describe the value is a gauge and only contains currently active requests.

francoisferrand replied:
Bad copy-paste indeed, good catch! 👍

williamlardier commented on lib/utilities/monitoringHandler.js:
```diff
@@ -169,21 +179,21 @@ function routeHandler(req, res, log, cb) {
 function monitoringHandler(clientIP, req, res, log) {
     function monitoringEndHandler(err, results) {
-        writeResponse(res, err, log, results, error => {
+        writeResponse(res, err, results, error => {
```
Is this function called with a `results` parameter that is not equal to `[]`? The `routeHandler` function never calls this function if there is no error.

francoisferrand replied:
correct, I don't think it is really used at the moment.
I don't know why it's there, but there may be a reason why it was added, i'd prefer not rushing to clean this thing..

francoisferrand committed:
fixup! Add http active requests metrics

francoisferrand committed:
Rename label "route" to "action"

It avoids the confusion with the actual route, and will be easier to
understand when looking at the metrics.

Issue: CLDSRV-63

francoisferrand committed:
Use high resolution clock to fix flacky tests

Issue: CLDSRV-63

francoisferrand commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-63 contains:


* `8.3.1`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.3.2`


Please check the `Fix Version/s` of CLDSRV-63, or the target
branch of this pull request.




*The following options are set:* **approve**

francoisferrand commented:
fixed version in ticket

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.3`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/8.2`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.3`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/8.2`



Please check the status of the associated issue CLDSRV-63.

Goodbye francoisferrand.
